### DISCUSSION TITLE
rendering raw rgb video data

### DISCUSSION THREAD

**Discussion by Mdbg** [emilk/egui#811] [Q&A]
what is the most efficient way to do that?

```
let pixels: Vec<_> = pixels
                .chunks_exact(4)
                .map(|p| egui::Color32::from_rgba_unmultiplied(p[0], p[1], p[2], p[3]))
                .collect();

let texture = frame
                .tex_allocator()
                .alloc_srgba_premultiplied(size, &pixels);
```

this way is too slow.

**Comment by emilk** [comment]
Ideally you want a video-decoder that decodes straight into video memory. If you don't have that, then avoiding the `from_rgba_unmultiplied` call and just padding in the raw pixels is a good optimization (unless you have an alpha channel in the video of course). You could/should also consider doing as much of the work as possible in a background thread.

**Reply by Mdbg** [reply]
As far as i konw, i can do it with SrgbTexture2d and register_native_texture.
I'm using the eframe::run_native in my main function , and alloc a SrgbTexture2d need the opengl context glium::Display, but the glium::Display is invisiable out of eframe::run_native function. Is there graceful way to do that ?

**Reply by ruabmbua** [reply]
I also want to display some video, and I am also missing access to the backend with eframe. Is it possible to expose glium / glow inside the run_native?

**Reply by emilk** [reply]
Not currently. You best bet is to instead use `egui_glow` directly.

**Reply by morajabi** [reply]
what's the latest recommend way to render decoded frames on the egui window?
